### ISSUE TITLE
Strange spikes in users online count

### ISSUE THREAD

**Issue by samim23** [tmeasday/meteor-presence#17]
i´m using presence in a production app over at hyperboring.com.
noticing strange spikes at time in the online users count. 3 users are online but it will show 50-100 for 3-4 seconds and then go gradually back down.
Today the entire server died and when i looked at the logs, it was presence killing it.
Could this be related somehow? any thoughts are super welcome!

here´s my error log http://snipt.org/Bnv7

**Comment by dburles** [comment]
Hey @samim23 I've come across this issue a couple of times, right now I'm trying to see if I can find a way to reproduce it. I'm also wanting to determine if this issue has been introduced in some of the later changes that have been made. Seems like @gabrielhpugliese has a similar issue: #12

**Comment by samim23** [comment]
ok cool. what is your recommendation for the mean time? i´m thinking about removing presence for now.

**Comment by dburles** [comment]
You could try downgrade to an earlier version, perhaps e59a12739685972c970f70956c4d3e78c31818ec
you'll have to manually install the package to do this, just clone presence into packages.

**Comment by gabrielhpugliese** [comment]
It's hard to debug. I tried to debug localhost and coderstv.com with 50 chrome tabs, but couldn't reproduce. Do not know where to console.log either.

**Comment by gabrielhpugliese** [comment]
Hey, I've built a package just for this purpose. I do the interval on
client and insert/update the presence on client too. I'm going to test it
thursday with lots of different people. If it works I'll make a package.

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
@gabrielhpugliese very much looking forward to your patch as especially the issues with counters going up to 100 for a second seems to be getting worse and worse with scale. i´ve been researching "not using a heart" https://github.com/tmeasday/meteor-presence/issues/10/ a bit and seems like there is a more fundamental fix to be had in conjunction with that, or i´m i off track?

**Comment by gabrielhpugliese** [comment]
Well, it worked like a charm for a few minutes. I think it needs more
improvement. I will open source it soon so you can help on improvements.
Or probably this is a NP-hard problem :P

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
NP hard problem? how come a quake server browser solved this efficiently back in 1999 already? ;-) Anyhow, my issues are getting worse and worse now i´m hitting 1k daily users, even suspecting its pretty bad for the overall performance, could that be true? I´m gonna put in a day next week and will try to merge some of these libraries into a working solution, will share if i succeed.

**Comment by gabrielhpugliese** [comment]
Well, the load on my application hosted on a single-core CPU 2GHz goes to
2. 95% of cpu is for main.js When I turn it off, it goes to bottom. I have
no idea why.
Next Meteor version will provide a package called Facts that will give for
free how many subscriptions are active. Idk if it's enough for you.

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by gabrielhpugliese** [comment]
Well, I just improved it. Instead of inserting things on client, defining
allow rules in server, I've created a method and it's WORKINGOMG with 0.15
load on my server.
Want to test ? https://github.com/gabrielhpugliese/meteor-users-counter

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
@gabrielhpugliese you are amazing! i´m going to give this a spin right now and test it in-depth - will give you feedback asap.

**Comment by gabrielhpugliese** [comment]
Forgot to say, it works with 0.6.4.1 only - what's the meteor version
you're using ?

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
Ah to bad, i´m stuck on meteor 0.6.6.3 for now until the new render engine reaches a stable release. Hope i will get to that soon and i´ll get to check out your fix.

**Comment by dburles** [comment]
Hey guys. I've just done a complete rewrite to make use of livedata. Would be awesome if you could try it out and let me know how it works for you. You can grab it from the livedata branch.

**Comment by gabrielhpugliese** [comment]
Nice David, gonna try it :) Thanks

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
@gabrielhpugliese thanks!, interested to hear how you go.

**Comment by samim23** [comment]
@dburles  very cool stuff, thanks! I´ve been testing on my localserver and so far works like a charm! haven´t performance monitored fully yet but looking better off the bat. Might give it a shoot on the production server soon, anything to keep in mind with the transition?

**Comment by dburles** [comment]
@samim23 Good to hear! the main differences I can think of are that A. the _id coming from the presences collection now is tied directly to the livedata session id, so these will drop on hot code reload, where previously it would reattach the session to the existing presence _id. and B. clients not connected through a socket, (long polling) take a while to drop off when disconnecting, not sure of the exact duration, but somewhere around 1 minute. I believe though that this may have already been an issue with the previous version. Anyway it's probably useful to be aware of this limitation.

**Comment by gabrielhpugliese** [comment]
Will this work on 6.4.1 ?

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
@gabrielhpugliese I don't think so as it's relying on upsert which was introduced with 0.6.6

**Comment by gabrielhpugliese** [comment]
Okay :(

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
If you're feeling adventurous, it should work if you modify the upsert to the typical, check if exists, then run either an update or insert.

**Comment by gabrielhpugliese** [comment]
Or maybe update my project to 0.6.6 :P I'm postponing it forever. My
smart.json have a lot of git tags of previous versions.

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
Yeah updating the project would be a good idea :)

**Comment by samim23** [comment]
@dburles i´m running your new version on the production server now at https://hyperboring.com. So far works perfect and no strange user counter raises anymore. Performance is much better to, with memory consumption and CPU time both down by allot. Nice job, this is great!

**Comment by mizzao** [comment]
You guys may want to check out https://github.com/mizzao/meteor-user-status, which I've used in many projects and seems to be pretty stable and straightforward.

**Comment by gabrielhpugliese** [comment]
I've never used it because on docs it does not mention the possibility to
add custom vars to the presence doc. Is it possible ?

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by mizzao** [comment]
@gabrielhpugliese I see - I always just assumed that anyone who would want to add anything on the client side would just add it to the `profile` field of a user doc, and that would be published/subscribed as normal. It didn't make sense to me to include an extra collection for this because presence information is already associated with each user (versus sessions, for example.)

**Comment by gabrielhpugliese** [comment]
For example, I need to know where an user is at the moment. I think it's
perfectly reasonable to stay on the presence doc, not on profile. It's
something ephemeral.

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
@dburles i´ve been testing your new code for an evening and works nicely, except for "ghost" users that get stuck in the system for hours even though nobody is there. happens only sporadically, only reboot seems to work as a fix.

**Comment by samim23** [comment]
@gabrielhpugliese the missing extra fields (and coffeescript ;-) ) was the main reason i decided against user-status. But i did not think it through it seems, as your solution described here makes perfect sense.

**Comment by gabrielhpugliese** [comment]
Yeah coffeescript +1

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by samim23** [comment]
@mizzao @gabrielhpugliese i´ve made a quick translation of the meteor-user-status / user_status.coffee from coffee into js: https://gist.github.com/samim23/7785333 not tested yet. @gabrielhpugliese solution for custom fields (querying the user profile object) makes sense so will try that, as @dburles newest fix unfortunately still produces strange results on the production server. @mizzao is it possible to track non logged in users to?

**Comment by mizzao** [comment]
@samim23 It looks you got the inverted message from @gabrielhpugliese, who said he liked coffeescript but didn't want to keep presence data around in `profile`. You might also want to update your conversion to js since I just made some changes as well :). Also, you could definitely track both logged-in and non-logged-in users with their session/socket id.

**Comment by mizzao** [comment]
@dburles and @samim23 I just realized the ghost users may be due to an incorrect use of upsert. See the latest commits on user-status.

**Comment by gabrielhpugliese** [comment]
No, I do not like coffee :(

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by gabrielhpugliese** [comment]
I think we should move this conversation to another place, like meteor-talk

Gabriel Pugliese
CodersTV.com
@coderstv

On Wed, Dec 4, 2013 at 2:42 PM, Gabriel Pugliese <gabrielhpugliese@gmail.com

> wrote:
> 
> No, I do not like coffee :(
> 
> Gabriel Pugliese
> CodersTV.com
> @coderstv
>

**Comment by samim23** [comment]
@mizzao tracking non-logged-in users with their session/socket id is clear, question is more where to store the additional field data for those as they have no profile. Of course could open a parallel collection for that but seems messy.
@dburles any clues what could cause those ghosts? I´m tempted to empty the collection serverside on a interval for now but seems hacky.

**Comment by dburles** [comment]
hey @samim23 I think the issue could have been caused by the upsert as @mizzao mentioned,  I realised it isn't actually required.  the updates are over on https://github.com/dburles/meteor-presence

**Comment by mizzao** [comment]
@samim23 I'm curious if you see ghosts when using user-status too. Might be the way you are using it, mind giving it a try?

@dburles you definitely need the upsert; a user logging in and out and back in, etc will have the same session id and you will get duplicate key errors.

**Comment by dburles** [comment]
oh yeah of course, argh. too early in the morning :)

**Comment by mizzao** [comment]
@dburles Take a look at https://github.com/meteor/meteor/issues/1552 to make sure you are using upsert correctly; if you don't include `$set` then it will just ignore the key you entered, possibly causing the ghosting.

**Comment by dburles** [comment]
@mizzao thanks. looking into it more, i can cause ghost users to appear by refreshing the browser rapidly while logged in. also, strangely the socket close event fires twice on the same id for logged in users.

**Comment by dburles** [comment]
Hey so i'm pretty confident i've now corrected the ghosting issue, the one i mentioned above. So assuming that this was causing the same problems that you were seeing @samim23 hopefully it will now no longer happen! It would be great if you could try it out and let me know if there are any issues @samim23. Make sure you grab it from here: https://github.com/dburles/meteor-presence

**Comment by samim23** [comment]
@dburles super cool! i´ve just uploaded your new version to my production server. Not much testing yet but will give you proper feedback soon.

**Comment by dburles** [comment]
thanks, appreciated!

**Comment by samim23** [comment]
@dburles So i´ve done some testing today, two issues: Ghosts still can appear even though far less frequent. Unfortunately it stopped tracking non-logged in users completely which is bad for my usecase.

**Comment by dburles** [comment]
@samim23 hey yeah sorry about that, I was doing a lot of testing with logged in users. I need to figure out how to write tests for this! I've patched it up now so just grab the code again from dburles/presence. If you don't mind, when you are seeing ghost users, could you show me what they appear like in the collection?

**Comment by samim23** [comment]
@dburles i´ve just uploaded your most recent patch but now in my server stats the project average response time went from somewhere around 100ms to over a value of 150000ms. Evil ! As i did not change much else, i´m asking myself if this could be related to your update at all? I´ll look into making a snapshot of my collection asap and sending it over.

**Comment by dburles** [comment]
Hmm I'm not sure how that correlates with the changes made. Wonder if
something else has gone askew. Very strange. Possible to  trace down the
problem?

**Comment by samim23** [comment]
I apologize as it seems to be just a snafu random correlation with my host going insane exactly after my upload. Now looking into making a custom output page, showing everything in the presence collection, will share that URL with you asap.

**Comment by gabrielhpugliese** [comment]
So it's nice to use now ?

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
I believe so! Give it a try, some real world testing is good :)

**Comment by dburles** [comment]
@gabrielhpugliese if you want to try it, make sure you grab it from https://github.com/dburles/meteor-presence

**Comment by gabrielhpugliese** [comment]
Do you mean https://github.com/dburles/meteor-presence/ ?

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
@gabrielhpugliese yep

**Comment by samim23** [comment]
@dburles did a intensive test with around 30 users over the course of a day. The ghosting issue is indeed still present, user will show up in the presence collection even though nobody is there. An additional new issue is, that sometimes users who are online in a room, wont be counted at all - leading to to low user count. Let me know how i can help to resolve these issues.

**Comment by dburles** [comment]
Would it be possible to pinpoint any areas that may trigger the ghost users? HCR? login/logout? etc
The ghost users within the collection, do they have a userId? and also when you say that the users are not counted, this could potentially also be an issue with your subscription/publication? I'm not sure. Any info you can dig up would be really helpful. Thanks :)

**Comment by samim23** [comment]
- The ghost users do have userIDs (I´m checking from the browser console) and look normal except that not all users in the the collection have a "state" field (where the "online" and my "roomID" are stored).
- Really hard to pinpoint trigger areas as i´m testing with a live service.
- Users not being counted is only occurring since the recent updates of presence & i didn´t change the pub/sub´s. I´ll be inside a room and it will show 0 users online, funny business ;)
- Will try to dig up more infos, not so easy ;-)

**Comment by samim23** [comment]
Quick update on the "user is in room but counter shows 0" issue. It seems like this occurs after a client is inactive for a while but remains in room/browser open/computer not sleeping. The disappearing users still show up in the overall presences counter though, just the field data gets lost. Not sure if this helps.

**Comment by gabrielhpugliese** [comment]
That is really bad for me since my app is a platform for long videos :(

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
Hey Gabriel there are some changes coming in the next version of meteor
that should allow us to track user sessions in a much nicer way, so I'm
going to give that a try soon.

**Comment by gabrielhpugliese** [comment]
Yeah. I'm going to stick with my own method. I just upgraded to 0.6.6.3. If
you want to try here's the link:
https://github.com/gabrielhpugliese/meteor-users-counter

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by dburles** [comment]
Yeah that's cool. The goal here is to not have to rely on polling so user
presence changes are realtime. Hopefully this new feature may solve these
bugs

**Comment by gabrielhpugliese** [comment]
Would save our lives :)

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by mizzao** [comment]
@samim23 @gabrielhpugliese and @dburles In [meteor-user-status](https://github.com/mizzao/meteor-user-status) am also (rarely) observing imaginary users being online after long periods of inactivity. It seems that the `socket.on("close", ...)` event must not be triggered.

Meteor 0.6.7 seems to give us the option to get out of this mess. Unfortunately, it doesn't keep track of IP addresses, yet. I think I will submit a PR for that :)

**Comment by gabrielhpugliese** [comment]
My users-counter package is not presenting ghost users :)

Gabriel Pugliese
CodersTV.com
@coderstv

**Comment by mizzao** [comment]
I just had a moment of clarity about this bug while washing dishes.

At least some of the problem in this package and also in meteor-user-status is the fact that the SockJS close event isn't fired when people sleep their computers, at least not for several hours. Sometimes it just isn't fired ever. This may be due to an older version of SockJS that Meteor is using or just something that isn't fixed yet, but I'm pretty sure this is the issue that's causing these ghost users.

**Comment by samim23** [comment]
@mizzao makes sense, thanks for the pointer. Has anybody had the chance looking into the new meteor "facts" package? might solve our issues

**Comment by dburles** [comment]
hey guys, i've decided to add a heartbeat here to deal with sleeping/unclosed sockets by maintaining a server side list of connected clients alongside a tick method.

I currently don't have any real-world apps to properly test this over time with so any help with that is really appreciated, thanks.

The updated presence package can be found over here: https://github.com/dburles/meteor-presence
